File: code/c++/REVC.cpp
#include <iostream>
#include <fstream>
#include <map>
#include <string>

class REVCSolution
{
private:
    std::string dna_seq, comple_seq, revc;

public:
    // lendo a sequência de DNA
    void read_dna_seq()
    {
        std::ifstream input_data("C:/Programming/Rosalind/input/rosalind_revc.txt");

        if (input_data.is_open())
        {
            std::getline(input_data, dna_seq, '\n');

            input_data.close();
        }
        else
        {
            std::cerr << "Não foi possível abrir o arquivo. " << std::endl;
        }
    }

    // inverter a sequência do complemento de dna
    std::string reverse(std::string sequence)
    {
        std::string newSequence;
        for (int i = sequence.length() - 1; i >= 0; --i)
        {
            newSequence += sequence[i];
        }
        return newSequence;
    }

    // pegar o complemento da sequência de dna
    void set_complement()
    {
        std::map<char, char> complement = {{'A', 'T'}, {'T', 'A'}, {'C', 'G'}, {'G', 'C'}};

        for (auto base : dna_seq)
        {
            comple_seq += complement[base];
        }
    }

    void get_reverse_complement()
    {
        read_dna_seq();
        set_complement();
        revc = reverse(comple_seq);

        std::ofstream output_data("C:/Programming/Rosalind/output/REVC.txt");

        if (output_data.is_open())
        {
            output_data << revc;

            output_data.close();
        }
        else
        {
            std::cerr << "Não foi possível abrir o arquivo. " << std::endl;
        }
    }
};

int main()
{
    REVCSolution solution;
    solution.get_reverse_complement();

    return 0;
}

File: code/python/REVC.py
import os


class REVCSolution:
    def __init__(self):
        self.dna = None
        self.dna_reverse_compl_seq = ""
        self.read_dna_seq()

    # lendo a sequência do dna
    def read_dna_seq(self):
        try:
            path = os.path.realpath("../../input/rosalind_revc.txt")
            input_data = open(path, 'r')
            self.dna = input_data.readline().strip()
            input_data.close()

        except IOError:
            print("Erro na leitura do arquivo!")

    # inverter sequência
    def reverse(self, complement):
        reverse_comp = ""
        for i in range(len(complement)-1, -1, -1):
            reverse_comp += complement[i]
        return reverse_comp

    # definir complemento
    def set_complement(self):
        pairs = {'A': 'T', 'T': 'A', 'C': 'G', 'G': 'C'}

        dna_complement_seq = ""
        for base in self.dna:
            dna_complement_seq += pairs[base]

        self.dna_reverse_compl_seq = self.reverse(dna_complement_seq)

    # gerando a saída
    def get_reverse_complement(self):
        self.set_complement()

        try:
            path = os.path.realpath("../../output/REVC.txt")
            output_data = open(path, 'w')
            output_data.write(self.dna_reverse_compl_seq)
            output_data.close()

        except IOError:
            print("Erro na escrita do arquivo!")


if __name__ == "__main__":
    solution = REVCSolution()
    solution.get_reverse_complement()
